### PR TITLE
feat(apm) Expose transaction start & end timestamp

### PR DESCRIPTION
**Pull request by markstory**
On transaction events expose both the start and end timestamps. This will allow us to correctly display the transaction span length.

I've also added the transaction sample from my now closed tracing pull request. Having a sample transaction event should be useful when we have to do more endpoint and acceptance tests around transaction events.

Refs SEN-799

### CODE REVIEW

File: src/sentry/api/serializers/models/event.py [hunk 1]
```diff
@@ -285,6 +285,11 @@ def serialize(self, obj, attrs, user):
                 'tags': tags_meta,
             },
         }
+        # transaction events have start and end-times that are
+        # timestamp floats.
+        if obj.get_event_type() == 'transaction':
+            d['startTimestamp'] = obj.data.get('start_timestamp')
+            d['timestamp'] = obj.data.get('timestamp')
```
Comment (fpacifici) [inline src/sentry/api/serializers/models/event.py]: I think this ends up being the same as obj.datetime and the field dateCreated that is populated at line 272. Is there a reason to add a new one ?

**Comment by dashed** [comment]
Is it possible to have a dedicated transaction object?

Something like

```
{
    // ...
    
    "transaction": {
        "name": "/country_by_code/",
        "start_timestamp": 1562681591.0,
        "end_timestamp": 1562681591.0,
    },

    // left alone for other parts of the UI
    "timestamp": 1562681591.0,
}
```

File: src/sentry/api/serializers/models/event.py (hunk 1 above)
Comment (markstory) [reply src/sentry/api/serializers/models/event.py]: The `dateCreated` field is a string and lacks the ms precision.

**Comment by markstory** [comment]
@dashed I'm reluctant to overload `transaction` as it means something different in the application internals. I could however add the name, start & end timestamps to `metadata` as that is a flex slot that varies by event type.

**Comment by dashed** [comment]
@markstory Yea the `metadata` slot may be suitable. As long as the transaction attributes are grouped together somewhere 😄 .

I feel like attributes `start_timestamp` don't really belong at the top level.

**Comment by markstory** [comment]
I've moved the start & end timestamps into the `TransactionEvent` metadata value. This resolves the structural issues in the API response, but has an impact on storage.

With timestamps in the metadata, the generated `message` that ends up in clickhouse contains both timestamps. An example of this is `1563298402.26 1563298402.24 /cities/`

**Comment by dashed** [comment]
I don't think we want the message to be like that 😬

File: src/sentry/api/serializers/models/event.py (hunk 1 above)
Comment (fpacifici) [reply src/sentry/api/serializers/models/event.py]: then, since this is going to be there only for transacitons, I would advise to rename the two fields in such a way that make it clear what they mean like:
startTimestamp
endTimestamp

So people would not have to look into the code to understand what timestamp means in the context of the transaction.

**Comment by markstory** [comment]
@dashed I've done some refactoring in EventManager to move the message generation into the event types. This change will prevent the timestamps from being indexed in the message.

@fpacifici Do you think the timestamp names need to change if they are located in `metadata`? I'd like to avoid having different names between the transaction event and the spans if possible.

File: src/sentry/eventtypes/transaction.py [hunk 2]
```diff
@@ -13,11 +13,21 @@ def get_metadata(self, data):
...
     def get_location(self, metadata):
         return metadata['location']
+
+    def build_search_message(self, message, metadata):
+        # Don't include timestamps in the search message.
+        message_data = {
+            'title': metadata['title'],
+            'location': metadata['location'],
+        }
+        return super(TransactionEvent, self).build_search_message(message, message_data)
```
Comment (dashed) [inline src/sentry/eventtypes/transaction.py]: @markstory that's a pretty good refactor 👍

**Comment by dashed** [comment]
This looks pretty good to me. 👏 

I'll leave the implementation details to @fpacifici since I don't have strong opinions on that front.

**Comment by fpacifici** [comment]
> Do you think the timestamp names need to change if they are located in metadata? I'd like to avoid having different names between the transaction event and the spans if possible.

meh .... ok. If we are end up being inconsistent either way feel free to ignore my comment.

File: src/sentry/eventtypes/base.py [hunk 3]
```diff
@@ -21,6 +23,13 @@ def get_title(self, metadata):
     def get_location(self, metadata):
         return None
 
+    def build_search_message(self, message, metadata):
```
Comment (fpacifici) [inline src/sentry/eventtypes/base.py]: nit, please use the same naming convention of the other methods. get_search_message.

But most important, the fact you are passing in the event message to this method makes me think this may not be the right place to add this. I would get a review from someone who is more familiar with the goal of the eventtypes data structure.

File: src/sentry/utils/samples.py [hunk 4]
```diff
@@ -139,7 +139,7 @@ def load_data(platform, default=None, sample_name=None):
         return
 
     data = CanonicalKeyDict(data)
-    if platform in ('csp', 'hkpk', 'expectct', 'expectstaple'):
+    if platform in ('transaction', 'csp', 'hkpk', 'expectct', 'expectstaple'):
```
Comment (fpacifici) [inline src/sentry/utils/samples.py]: Why is this change needed ? It was not here in the previous version if I remember correctly

Comment (markstory) [reply src/sentry/utils/samples.py]: Without this the sample transaction event gets data added to the `extra`, `modules` and `request` interfaces which are generally not going to be present on a transaction event payload.

File: src/sentry/eventtypes/base.py (hunk 3 above)
Comment (markstory) [reply src/sentry/eventtypes/base.py]: That is a good point. I'll consolidate the message generation further so that it is less spread out across `EventManager` and the eventtypes.

**Comment by markstory** [comment]
I've removed the `metadata` attribute. I talked with @jan-auer and event metadata is intended to capture data that needs to be persisted to the 'group'. Since transaction events don't have groups using this doesn't follow the original intent.

I've gone back to exposing timestamps on the root level of the event and using `startTimestamp` and `endTimestamp` to be more clear. While this creates some dissimilarity with how spans represent ending timestamp, I'm open to renaming those properties as well.

I've also shifted some of the properties that will only appear on error events around so that those keys are never populated on transaction events.

**Review by dashed — APPROVED** [review]
Looks good to me. 👍 

I'll let either @fpacifici or Vienna give their feedback (cc @jan-auer )

File: src/sentry/api/serializers/models/event.py [hunk 5]
```diff
@@ -285,8 +280,34 @@ def serialize(self, obj, attrs, user):
...
             },
         }
+        # Serialize attributes that are specific to different types of events.
+        if obj.get_event_type() == 'transaction':
+            d.update(self.serialize_transaction_attrs(attrs, obj))
+        else:
+            d.update(self.serialize_error_attrs(attrs, obj))
         return d
 
+    def serialize_transaction_attrs(self, attrs, obj):
```
Comment (fpacifici) [inline src/sentry/api/serializers/models/event.py]: I guess this and serialize_error_attrs should be "private" 
__serialize_error_attrs

File: src/sentry/api/serializers/models/event.py [hunk 6]
```diff
@@ -269,11 +267,8 @@ def serialize(self, obj, attrs, user):
             'metadata': obj.get_event_metadata(),
             'tags': tags,
             'platform': obj.platform,
-            'dateCreated': obj.datetime,
```
Comment (fpacifici) [inline src/sentry/api/serializers/models/event.py]: I noticed a few dependencies on dateCreated 

>src/sentry/static/sentry/app/views/organizationEventsV2/eventModalContent.jsx:          date={getDynamicText({value: event.dateCreated, fixed: 'Dummy timestamp'})}
>src/sentry/static/sentry/app/views/organizationEventsV2/modalLineGraph.jsx:  const eventTime = +new Date(currentEvent.dateCreated);
>src/sentry/static/sentry/app/views/organizationGroupDetails/eventToolbar.jsx:    const dateCreated = moment(evt.dateCreated);

Are these going to be ok if you remove the field? I suspect there is no acceptance test that go through transaction event.

Comment (markstory) [reply src/sentry/api/serializers/models/event.py]: Good finds. The modal graph and content component do need to be updated. The GroupDetailsToolbar should be fine as transaction events won't have groups.

File: src/sentry/utils/samples.py (hunk 4 above)
Comment (markstory) [reply src/sentry/utils/samples.py]: The variable name `platform` is an awkward name as it is used to locate a sample/fixture file in `src/sentry/data/samples`. The files there are a mixture of both sample SDK events and special scenarios like `empty-exception` or `python-omittedbody`.

**Comment by fpacifici** [comment]
I am ok with this now. @dashed you may want to have a look at the UI change.
